util/tls/tls_socket.cc

```diff
@@ -505,13 +507,138 @@ void TlsSocket::OnRecv(const RecvNotification& rn, const OnRecvCb& recv_cb) {
 }
 
 io::Result<size_t> TlsSocket::TrySend(io::Bytes buf) {
-  LOG(DFATAL) << "Not implemented";
-  return 0;
+  iovec vec[1];
+  vec[0].iov_base = const_cast<uint8_t*>(buf.data());
+  vec[0].iov_len = buf.size();
+  return TrySend(vec, 1);
 }
 
 io::Result<size_t> TlsSocket::TrySend(const iovec* v, uint32_t len) {
-  LOG(DFATAL) << "Not implemented";
-  return 0;
+  if (IsEmptyIovec(v, len)) {
+    DCHECK(false) << "TrySend with empty iovec";
+    return 0;  // nothing to send (POSIX allows zero-length writes)
+  }
+  if ((state_ & WRITE_IN_PROGRESS) != 0) {
+    // Another fiber is currently writing, we cannot safely proceed
+    DVSOCK(3) << "TrySend blocked: WRITE_IN_PROGRESS detected";
+    return make_unexpected(make_error_code(errc::resource_unavailable_try_again));
+  }
+  bool read_in_progress{(state_ & READ_IN_PROGRESS) != 0};
+  size_t total_bytes_sent{};
+  std::error_code returned_status{};
+  // We make a local mutable copy of the iovec descriptors because AdvanceIovec
+  // modifies them (adjusting base pointers and lengths) to track partial writes.
+  // The input array 'v' is const and belongs to the caller, so we cannot modify it.
+  static constexpr size_t kMaxStackIovecs = 16;
+  uint32_t curr_iovec_len{len};
+  absl::InlinedVector<iovec, kMaxStackIovecs> curr_iov(curr_iovec_len);
+  iovec* iov_cursor = curr_iov.data();
+  std::memcpy(iov_cursor, v, curr_iovec_len * sizeof(iovec));
+
+  while ((curr_iovec_len > 0) || (engine_->OutputPending() > 0)) {
+    // 1. Flush into the upstream socket any pending output from the engine output buffer before
+    // pushing more data to the engine from the user. These might be bytes from previous call.
+    if (engine_->OutputPending() > 0) {
+      auto output_buf{engine_->PeekOutputBuf()};
+      DCHECK(!output_buf.empty());
+      auto send_result{next_sock_->TrySend(output_buf)};
+      if (send_result) {
+        CHECK_LE(*send_result, output_buf.size());
+        engine_->ConsumeOutputBuf(*send_result);
+        DVSOCK(3) << "Flushed " << *send_result << " bytes to upstream";
+        if ((*send_result) < output_buf.size()) {  // case 1.A: partial write
+          // upstream socket is full - try again later
+          returned_status = make_error_code(errc::resource_unavailable_try_again);
```

Comment:
1. Bug. what ensures that the write will be flushed when `next_sock_` becomes available?
2. `make_error_code(errc::resource_unavailable_try_again);` is wrong as you already consumed some of the input data and it was copied to ssl engine. From a caller perspective - they need to retry the entire operation, so they will try to push the same data again